src/Workspaces/CSharp/Portable/Simplification/Reducers/CSharpNullableAnnotationReducer.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CSharp.Simplification
+{
+    internal sealed partial class CSharpNullableAnnotationReducer : AbstractCSharpReducer
+    {
+        private static readonly ObjectPool<IReductionRewriter> s_pool = new ObjectPool<IReductionRewriter>(
+            () => new Rewriter(s_pool));
+
+        public CSharpNullableAnnotationReducer() : base(s_pool)
+        {
+        }
+
+        private static readonly Func<NullableTypeSyntax, SemanticModel, OptionSet, CancellationToken, SyntaxNode> s_simplifyNullableType = SimplifyNullableType;
+
+        private static SyntaxNode SimplifyNullableType(
+            NullableTypeSyntax node,
+            SemanticModel semanticModel,
+            OptionSet optionSet,
+            CancellationToken cancellationToken)
+        {
+            // If annotations are enabled, there's no further simplification to do
+            var context = semanticModel.GetNullableContext(node.Span.End);
+
+            // Work around https://github.com/dotnet/roslyn/issues/37809
+            if (semanticModel.IsSpeculativeSemanticModel && context.AnnotationsInherited())
+            {
+                // Work around bug where GetNullableContext() on a speculative model doesn't inherit automatically
+                context = semanticModel.ParentModel.GetNullableContext(semanticModel.OriginalPositionForSpeculation);
+            }
+
+            if (context.AnnotationsEnabled())
+            {
+                return node;
+            }
+
+            // If it's not a reference type, also leave the ?
+            var type = semanticModel.GetTypeInfo(node, cancellationToken).Type;
+            if (type == null || !type.IsReferenceType)
```

Comment:
what sort type is an error type?  i think it woudl be good to not have us reduce in that case since we're not sure what's going on.  i.e.  if i write `SomeValueType?` and `SomeValueType` doesn't exist here, probably best to not touch things.

Reply:
I could actually imagine us intentionally reducing in those cases: say you have something like:

```
#nullable disable
Blah x = Foo() ?? Bar()
```

where Blah is an error type since it's unresolved. If you generate method on Foo(), should it be `Blah Foo()` or `Blah? Foo()`? The type inferrer will return `Blah?` since it knows the intent is to be nullable, but the resulting ? is only useful it was going to be a struct. My gut is saying people _generally_ do classes and only structs "when necessary"...?

Reply:
Either way we absolutely should capture this test behavior in a test. I could be convinced in either direction the right behavior though!

Reply:
Fair enough! I'm ok with that reasoning :)

Reply:
Oh this is a bit funkier than I expected: if you have `ErrorType?` the compiler assumes it must be a struct and thus does produce a `Nullable<ErrorType>`. So I can either:

1. revise this check to remove question marks from `Nullable<ErrorType>`s unilaterally, but it worries me we might end up dropping the question mark then when we did want it, like say generating from `Blah? x = Foo()`, or:
2. add some extra bit we carry along that says this is "maybe maybenull" and then simplify on it.
3. say the type inferrer doesn't go adding nullable annotations to error types.

I'll do 3. 2 would be cool but would require some unhappiness of carrying that extra data along.

Reply:
I'm going to do 3 and file a bug to do 2. This PR has gotten big enough as it is and I don't want to result in nitpicking about this corner case to hold up the whole thing.

Reply:
https://github.com/dotnet/roslyn/issues/37852 will track the better fix up.